### PR TITLE
Add baseline Quad Z stepper support

### PR DESCRIPTION
**Pull request by InsanityAutomation**
Following needs work :
~~Motion.cpp Function starting Line 1593
pins.h
pinsDebug.h
sensitivepins
G34~~

Add test with enough steppers

Add quad Z steppers as required for Voron functionality. Will still need more work of course but I wanted to get what I have up for review.

This should be pretty close. Ive done everything I can to verify this does not impact anything with current configurations. @sjasonsmith can you test this on youre triple Z machine? I dont have one here.

I have parts on the way to physically test this but personally I think this is close enough to merge with an experimental tag on the config option. Once I get the GTR1.0 board in, ill test that and add a test for quad z with it

### CODE REVIEW

**Comment by sjasonsmith** [comment]
Will there eventually be a new G34 algorithm for this, or will it use the simple iterative approach that is used for dual Z? I haven't looked at how Klipper or RepRap do it, to understand if there is a more efficient way to handle four steppers.

**Comment by InsanityAutomation** [comment]
Some of the guys on the voron team had told me they had already developed g34 changes for quad Gantry leveling. They were going to try and find the old code and if they can dig it out it will get implemented however they did it since that's been tested and proven in the past. If they can't find it next I will be going over to the klipper source which is iterative through the number of z steppers with a separate class called for voron machines explicitly.

**Comment by thinkyhead** [comment]
This should probably be directed to the `dev-2.1.x` branch rather than the bugfix branch.

**Comment by InsanityAutomation** [comment]
It is quite sizable so I don't disagree. I'll check how up to date it as and re point this after a few more tweaks when I get out of the day job tomorrow.

**Comment by InsanityAutomation** [comment]
Rebased and got this up to date. Have not moved anything to dev2.1 as its quite far behind 2.0.

**Comment by AnHardt** [comment]
Pardon.
I don't understand how this is able not to over constrain/bow/lock a plane bed.
Normally at a plane fixed to 4 points there is no way to move one.

**Comment by InsanityAutomation** [comment]
> Pardon.
> I don't understand how this is able not to over constrain/bow/lock a plane bed.
> Normally at a plane fixed to 4 points there is no way to move one.

The voron uses 4 belt driven z axis. Industrially I've also done it with pivoting brackets on each corner. The whole Gantry essentially floats. Bed itself is fixed.

**Comment by InsanityAutomation** [comment]
@AnHardt also worth noting this is already used in both RRF and klipper by the voron machines.

Edit : this is also useful on very large machines where current requirements force multiple drivers even if they cant physically level themselves due to constraints.

**Comment by InsanityAutomation** [comment]
received_1062429194105582
received_309587839928448
received_2830696820315063
received_448404235856277

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -2243,14 +2272,22 @@
     #define Z2_CHAIN_POS        -1
   #endif
 
-  #if AXIS_DRIVER_TYPE_Z3(L6470)
+#if AXIS_DRIVER_TYPE_Z3(L6470)
```
Comment (sjasonsmith) [inline Marlin/Configuration_adv.h]: This indent should not have changed. Likely refers to other configuration files as well.
This applies to all Configuration_adv.h files.

File: config/default/Configuration_adv.h [hunk 2]
```diff
@@ -2024,6 +2053,8 @@
   #define Y2_SLAVE_ADDRESS 0
   #define Z2_SLAVE_ADDRESS 0
   #define Z3_SLAVE_ADDRESS 0
+  #define Z4_SLAVE_ADDRESS 0
+  #define Z4_SLAVE_ADDRESS 0
```
Comment (sjasonsmith) [inline config/default/Configuration_adv.h]: `Z4_SLAVE_ADDRESS` is present twice. Only present in this file. Other `Configuration_adv.h` files look fine.

File: config/default/Configuration_adv.h [hunk 3]
```diff
@@ -2099,6 +2130,8 @@
   #define Z_HYBRID_THRESHOLD       3
   #define Z2_HYBRID_THRESHOLD      3
   #define Z3_HYBRID_THRESHOLD      3
+  #define Z4_HYBRID_THRESHOLD      3
+  #define Z4_HYBRID_THRESHOLD      3
```
Comment (sjasonsmith) [inline config/default/Configuration_adv.h]: Defined twice. Other files are fine.

File: config/default/Configuration_adv.h [hunk 4]
```diff
@@ -2308,7 +2349,7 @@
    *         I not present or I0 or I1 - X, Y, Z or E0
    *         I2 - X2, Y2, Z2 or E1
    *         I3 - Z3 or E3
-   *         I4 - E4
+   *         I4 - Z4 or E4 or Z4
```
Comment (sjasonsmith) [inline config/default/Configuration_adv.h]: This comment mentions Z4 twice. Other files are fine.

File: Marlin/src/gcode/calibrate/G34_M422.cpp [hunk 5]
```diff
@@ -65,7 +65,9 @@ constexpr xy_pos_t test_z_stepper_align_xy[] = Z_STEPPER_ALIGN_XY;
 #else
 
   static_assert(COUNT(test_z_stepper_align_xy) == Z_STEPPER_COUNT,
-    #if ENABLED(Z_TRIPLE_STEPPER_DRIVERS)
+    #if ENABLED(Z_QUAD_STEPPER_DRIVERS)
+      "Z_STEPPER_ALIGN_XY requires four {X,Y} entries (Z, Z2, Z3 and Z4)."
+    #elif ENABLED(Z_TRIPLE_STEPPER_DRIVERS)
       "Z_STEPPER_ALIGN_XY requires three {X,Y} entries (Z, Z2, and Z3)."
```
Comment (sjasonsmith) [inline Marlin/src/gcode/calibrate/G34_M422.cpp]: Mismatching Oxford commas before the `and`

File: Marlin/src/gcode/calibrate/M666.cpp [hunk 6]
```diff
@@ -57,9 +57,10 @@
    * M666: Set Dual Endstops offsets for X, Y, and/or Z.
    *       With no parameters report current offsets.
    *
-   * For Triple Z Endstops:
+   * For Triple / Quad Z Endstops:
    *   Set Z2 Only: M666 S2 Z<offset>
    *   Set Z3 Only: M666 S3 Z<offset>
+   *   Set Z4 Only: M666 S4 Z<offset>
    *      Set Both: M666 Z<offset>
    */
```
Comment (sjasonsmith) [inline Marlin/src/gcode/calibrate/M666.cpp]: `Set Both` should be be `Set All`.

File: Marlin/src/gcode/calibrate/M666.cpp [hunk 7]
```diff
@@ -69,7 +70,15 @@
...
-    #if ENABLED(Z_TRIPLE_ENDSTOPS)
+    #if ENABLED(Z_QUAD_ENDSTOPS)
+      if (parser.seenval('Z')) {
+        const float z_adj = parser.value_linear_units();
+        const int ind = parser.intval('S');
+        if (!ind || ind == 2) endstops.z2_endstop_adj = z_adj;
+        if (!ind || ind == 3) endstops.z3_endstop_adj = z_adj;
+        if (!ind || ind == 4) endstops.z4_endstop_adj = z_adj;
+      }
+    #elif ENABLED(Z_TRIPLE_ENDSTOPS)
```
Comment (sjasonsmith) [inline Marlin/src/gcode/calibrate/M666.cpp]: You could eliminate some duplication with this form:

```
    #if ANY(Z_TRIPLE_ENDSTOPS, Z_QUAD_ENDSTOPS)
      if (parser.seenval('Z')) {
        const float z_adj = parser.value_linear_units();
        const int ind = parser.intval('S');
        if (!ind || ind == 2) endstops.z2_endstop_adj = z_adj;
        if (!ind || ind == 3) endstops.z3_endstop_adj = z_adj;
        #if ENABLED(Z_QUAD_ENDSTOPS)
          if (!ind || ind == 4) endstops.z4_endstop_adj = z_adj;
        #endif
      }
```

File: Marlin/src/inc/SanityCheck.h [hunk 8]
```diff
@@ -509,22 +509,22 @@
...
-  #endif
+#elif BOTH(Z_TRIPLE_STEPPER_DRIVERS, Z_QUAD_STEPPER_DRIVERS) || BOTH(Z_DUAL_STEPPER_DRIVERS, Z_QUAD_STEPPER_DRIVERS) || BOTH(Z_DUAL_STEPPER_DRIVERS, Z_TRIPLE_STEPPER_DRIVERS)
+  #error "Please select only one Z_QUAD_STEPPER_DRIVERS, Z_TRIPLE_STEPPER_DRIVERS, or Z_DUAL_STEPPER_DRIVERS."
+#elif ENABLED(Z_DUAL_STEPPER_DRIVERS) && !(HAS_Z2_ENABLE && HAS_Z2_STEP && HAS_Z2_DIR)
+  #error "Z_DUAL_STEPPER_DRIVERS requires Z2 pins (and an extra E plug)."
 #elif ENABLED(Z_TRIPLE_STEPPER_DRIVERS) && !(HAS_Z2_ENABLE && HAS_Z2_STEP && HAS_Z2_DIR && HAS_Z3_ENABLE && HAS_Z3_STEP && HAS_Z3_DIR)
-  #error "Z_TRIPLE_STEPPER_DRIVERS requires Z3 pins (and two extra E plugs)."
+  #error "Z_TRIPLE_STEPPER_DRIVERS requires Z2 & Z3 pins (and two extra E plugs)."
+#elif ENABLED(Z_QUAD_STEPPER_DRIVERS) && !(HAS_Z2_ENABLE && HAS_Z2_STEP && HAS_Z2_DIR && HAS_Z3_ENABLE && HAS_Z3_STEP && HAS_Z3_DIR && HAS_Z4_ENABLE && HAS_Z4_STEP && HAS_Z4_DIR)
+  #error "Z_QUAD_STEPPER_DRIVERS requires Z2, Z3, Z4 pins (and three extra E plugs)."
```
Comment (sjasonsmith) [inline Marlin/src/inc/SanityCheck.h]: The `extra E plug` comments seem a bit strange to me.

I know that we try to find E plugs to auto-assign extra Z drivers to, but if that is the case, the user does not need to provide Z pins at all.

The comment indicates that they need Z pins _and_ E plugs, but in reality they need Z pins _or_ unused E drivers.

File: Marlin/src/module/configuration_store.cpp [hunk 9]
```diff
@@ -116,10 +116,10 @@
 
 #pragma pack(push, 1) // No padding between variables
 
-typedef struct { uint16_t X, Y, Z, X2, Y2, Z2, Z3, E0, E1, E2, E3, E4, E5; } tmc_stepper_current_t;
-typedef struct { uint32_t X, Y, Z, X2, Y2, Z2, Z3, E0, E1, E2, E3, E4, E5; } tmc_hybrid_threshold_t;
+typedef struct { uint16_t X, Y, Z, X2, Y2, Z2, Z3, Z4, E0, E1, E2, E3, E4, E5; } tmc_stepper_current_t;
+typedef struct { uint32_t X, Y, Z, X2, Y2, Z2, Z3, Z4, E0, E1, E2, E3, E4, E5; } tmc_hybrid_threshold_t;
```
Comment (sjasonsmith) [inline Marlin/src/module/configuration_store.cpp:124]: `EEPROM_VERSION` will need to be bumped with this change.

File: Marlin/src/module/endstops.cpp [hunk 10]
```diff
@@ -769,7 +811,9 @@ void Endstops::update() {
     }
     else { // Z +direction. Gantry up, bed down.
       #if HAS_Z_MAX || (Z_SPI_SENSORLESS && Z_HOME_DIR > 0)
-        #if ENABLED(Z_TRIPLE_ENDSTOPS)
+        #if ENABLED(Z_QUAD_ENDSTOPS)
+          PROCESS_TRIPLE_ENDSTOP(Z, Z2, Z3, Z4, MAX);
```
Comment (sjasonsmith) [inline Marlin/src/module/endstops.cpp]: Should be `PROCESS_QUAD_ENDSTOPS`

File: Marlin/src/module/endstops.h [hunk 11]
```diff
@@ -50,7 +51,10 @@ class Endstops {
       #if Z_MULTI_ENDSTOPS
         static float z2_endstop_adj;
       #endif
-      #if ENABLED(Z_TRIPLE_ENDSTOPS)
+      #if ANY(Z_TRIPLE_ENDSTOPS, Z_QUAD_ENDSTOPS)
+        static float z3_endstop_adj;
+      #endif
+      #if ENABLED(Z_QUAD_ENDSTOPS)
         static float z3_endstop_adj;
       #endif
```
Comment (sjasonsmith) [inline Marlin/src/module/endstops.h]: Should be `z4_endstop_adj`

File: Marlin/src/module/motion.cpp [hunk 12]
```diff
@@ -1628,6 +1633,52 @@ void homeaxis(const AxisEnum axis) {
...
+          // lock the second stepper for the final correction
+          (*lock[1])(true);
+          do_homing_move(axis, adj[2] - adj[1]);
+        }
+        else {
+          (*lock[2])(true);
+          do_homing_move(axis, adj[1] - adj[2]);
+          (*lock[1])(true);
+          do_homing_move(axis, adj[0] - adj[1]);
+        }
```
Comment (sjasonsmith) [inline Marlin/src/module/motion.cpp]: This algorithm looks simply duplicated from Triple Z. It will need changes to actually home to all four endstops.

**Review by sjasonsmith — COMMENTED** [review]
There are probably a lot of opportunities to do further cleanup to reduce duplication in these changes. Given the size of this PR, I think it makes more sense to ignore those for now, and treat that as an opportunity for a later round of Multi-Z cleanup.

**Comment by sjasonsmith** [comment]
For my review above I looked at every change in every non-example file. I looked at the configs in the Marlin folder and the default example, but not the others.

**Comment by sjasonsmith** [comment]
I also rebased my Triple-Z configuration onto this and verified it still compiles file. I haven't actually tested it yet. I will try to do that tomorrow.

**Comment by InsanityAutomation** [comment]
Awesome, I'll fix up the comments I botched and other little bits before then :)

File: Marlin/src/module/motion.cpp (hunk 12 above)
Comment (InsanityAutomation) [reply Marlin/src/module/motion.cpp]: By the look of it I got sidetracked halfway through this function and didnt finish it... Ill get this one tomorrow probably.

File: Marlin/src/inc/SanityCheck.h (hunk 8 above)
Comment (InsanityAutomation) [reply Marlin/src/inc/SanityCheck.h]: I maintained existing so ill let @thinkyhead determine if he wants to change the comments around.

**Comment by AnHardt** [comment]
Thanks.
Now i see. Looks as if, the joints of the gantry are that weak, that gantry will not be  plane without that much support points.

> also worth noting this is already used in both RRF and klipper by the voron machines.

That alone is no reason. There is conceptual poor hardware supported by all the firmwares. (Probes sending a trigger pulse, for example)

**Comment by sjasonsmith** [comment]
I tested with my Triple Z after commit 980b586663e4576aa850a97c4f3246219da9c735, and did not encounter any issues. I tested it both with and without known stepper positions.

**Comment by InsanityAutomation** [comment]
@sjasonsmith woohoo! So at minimum I failed to break anything :) hopefully with that last tweak up there we can look at something that may be merged soon then follow up as we can test various setups.

File: Marlin/src/inc/SanityCheck.h (hunk 8 above)
Comment (thinkyhead) [reply Marlin/src/inc/SanityCheck.h]: "(and/or three extra E plugs)"

**Comment by thinkyhead** [comment]
Rebased, squashed, and refactored Z multiple driver / endstop settings and related code. Use `git fetch origin` and `git reset --hard` as usual to get the latest changes. 

The CI tests are failing because various example configs are used during testing, and the configs are not yet up to date with this PR. I have a workaround, but I need to formalize it so it's easier for the CI script to grab the updated configs that apply to the PR being tested. For now I can just stick them in `Configurations/master` and re-run the tests.

**Comment by InsanityAutomation** [comment]
Sounds good! I have tracking on an 11 stepper @bigtreetech gtr1.0 board to test this with soon!

**Comment by thinkyhead** [comment]
> I have tracking on an 11 stepper … gtr1.0

Ours should arrive around the same day as yours.
It will be fun to get it to working in the E3D Toolchanger.

**Comment by InsanityAutomation** [comment]
Yup! And you know my other quad tool plan after that!

**Comment by thinkyhead** [comment]
This will be merged to `bugfix-2.0.x` pretty soon for inclusion in 2.0.2. At the moment I'm just running into a namespace issue (among others) w/r/t the L6470 handling. Should be sorted soon.

**Comment by thinkyhead** [comment]
It looks like your original commit added back some files that are obsolete. Can you isolate the changes you made to the "L6470" files? These have been renamed to L64XX so whatever changes you made to those files should be migrated to the new files.

**Comment by thinkyhead** [comment]
I think I got all the needed changes for L64XX. It should now pass testing.

**Pull request merged**